### DISCUSSION TITLE
Project Structure

### DISCUSSION THREAD

**Discussion by Morbidous** [MNES-Emulator/MNES#18] [Q&A] [answered]
I am going to create some new projects as I feel I need to. I might add Solution Folders. As mentioned before, the truly core projects should be named probably `Jamie.Core` etc.

For project structure I try to use what is called "hexagonal architecture". I don't know that it necessarily makes sense here or not. I will try to use DI in places, I hope you don't hate that. I think it would make a big difference for testing.

Generally I would structure a project similar to this:
 - Logic Library - As much dependency-free code as possible that relates to the project. Code that can be generalized gets pushed "up" into the core logic library instead. Has no project references besides the core logic library and maybe some other logic libraries (depends how your systems are arranged).
 - "Data" Library - Hides the dependency, whatever it is. Usually but not always the database. This has a reference only to the logic library (and maybe other stuff from "above" like other generalized data libraries and maybe in rare cases some other logic library.) In hexagonal arch. this is known as "driven" (it is the non-interface dependency).
 - "View" Library - This is the UI library. It has a reference to the logic library (and the general view library). This hides the dependency on the UI framework. You would have one of these for every UI framework you use. (For me this is often one for WinForms and one for web, for example.) In hexagonal arch. this is known as "driver" (the interface; the place where commands/messages into the system are generated).

So, in other words this looks like this: `Data --> Logic <-- View`

DI (dependency injection) is used to put it all together. DI uses what is call a "composition root" which is an important concept to know. This is basically the start of  your program where you construct all of the different parts, put them together, and run them. (Where the composition root is free to know about all dependencies, but the parts of the system don't know about the hard dependencies, everything is hidden behind interfaces.)

The program projects would have the comp root and almost nothing else. They would reference all of the Data libs, the Logic lib, and maybe one View lib (depending what you're doing) and put it all together.

(Actually you can have a program project that pulls data/logic/view from many different places and puts them together and runs them all at once in a much larger program that has access to multiple things. Although to have one "hexagon" (`Data --> Logic <-- View`) work together with another one requires creating an intermediate one that knows about both.)

(Note: "hexagonal"/"hexagon" is not important as a concept it's just a name; there is nothing to do with actual hexagons/six-sided shapes. The idea is a central dependency-free core with ports and adapters (its other name) used to communicate with the dependencies (the sides of the hexagon representing these interfaces to the dependencies where the core does not know about them but talks to them through the interface.))

And then, in addition to `Data --> Logic <-- View`, as said, below that I will have one or two program projects. For tests, I tend to separate the test project in two. I use the term "Testing" to indicate test-specific logic (code that helps you write tests, not the tests themselves). And then "Tests" for the project that actually has the tests. This lets a Tests project know about another Testing project so it can use test helper code from some other system without having to also have a reference to the tests from that system.

(An example of this that doesn't use databases might be Excel Interop. The "Data" library has the hard dependency to the Excel Interop assemblies but hides it behind interfaces. The Logic library has the dependency-free API that lets you do fun stuff with Excel. The View library has any UI-specific code that might need to exist. (Specifically UI code that needs to know about the Excel Logic library for some reason.)

The big thing here is that the View does not know about Excel at all, except from what is in the API in the Logic library. And the "Data" library does not even know what a UI is. Both Data and View know about "Excel" as defined in the dependency-free API; "Data" provides injections so that Logic can be used with or without "real" Excel behind it; View uses Logic without knowing if real Excel is really there or not.

(View is also allowed to know about Core View. Data is also allowed to know about Core Data, if there is one, but in this example there wouldn't be. For a database project there usually would be a small amount of code in a database-provider-specific general-use Data library.)

Just trying to "quickly" explain this so you have an idea what it looks like. This is not necessarily applicable to this project, but will probably be a little, and if/when I make some project structural changes you'll know what I'm thinking hopefully.

(For this project, there is little to no hard dependency, I think. The `MachineState` etc. is already dependency-free for the most part so there is no need for the "Data"/Driven side at least in the main system. But I may still find some things to inject and will likely try to hide certain components behind interfaces so that things can be tested in isolation.)

(Another thing to say is that my version of "hexagonal" might not be exactly the same as a more professional one. Really the main insight is the `Data --> Logic` part. Without hexagonal, very often a simple project would be structured like this: `Data <-- View`, where the View knows about the database. But then you have database code in your UI, which leads to a mess. (My first projects didn't even have Data/View separation which leads to even worse messes. The database code knows about the UI!) After `Data <-- View` the next logical step is `Data <-- Logic <-- View`, but still the UI knows about the database code, and the Logic library is just kind of in the way and doesn't do anything. Ah, but then you learn that Dependency Injection enables Dependency Inversion: literally inverting a dependency. This lets you reverse any dependency arrow that you want. So then you turn `Data <-- Logic <-- View` into `Data --> Logic <-- View` and now we have a dependency-free core and the UI does not know about the database and the database does not know about the UI. This is golden.)

**Comment by Morbidous** [comment]
I guess the file folder structure doesn't match the solution folder structure yet but that will involve moving all of the projects around.

**Comment by jamieyello** [comment, marked as answer]
Hey, btw, I've been gone for the last week because I literally have not had an hour of free time in the last week. It's been either 3 hours of sleep, or 20 hours to catch up. Two moves, and one move I had no help, so after 12 hours of a physically demanding job I was carrying furniture by myself for another 12, only to get no sleep and go right back to that job... I fell asleep while standing up at work. That should all be over now as now I'm on vacation.

Yeah you're the expert on these things, frankly. You've been telling me for years to have my own library.

So if it helps you while you're planning this out, I'm hoping to eventually add other emulator cores to the project, and possibly use Retroarch cores (I need to work out permissions, so I don't make anyone upset, of course). I believe atm this will all integrate Godot very tightly, and possibly exclusively, since I envision an environment where developers can create mods for emulator games that are written in a game engine environment.

This will probably veer development towards non-Nintendo games, since we all know how they feel about fan projects that get too ambitious and viral, but that's an issue for later.

But yeah I'm ok with you organizing things as you see fit.

**Reply by Morbidous** [reply]
(So you're a mover? How do you move furniture by yourself? I hired a moving company a couple years ago and there were three guys: the driver who was in charge, and two workers who followed his instructions. I can't imagine doing it alone; if it was doable by one person I wouldn't need to hire anyone in the first place. (I had already done everything I could do myself by myself though that turned out to be stupid since you have to pay them three hours and it only took them ~1.5 hours to do the rest.))

**Comment by jamieyello** [comment]
Just be sure to pull the most recent push I made, since it's a pretty big merge. I didn't leave things without some errors, the ini parser broke somewhere along the way, which might be a good excuse to find a more up to date library anyway.

**Reply by Morbidous** [reply]
yeah I try to do a pull + rebase every now and then to stay up to date. it seems like there is an issue with this whole process where we can't tell if/when the other person is working on what. I'm not sure how that works with larger projects with tons of devs. I guess they don't do so many large structural changes, or have a meeting about doing it before doing it or something like that.

**Reply by jamieyello** [reply]
Yeah there's always some friction working with other people. Pulling before you do any work and communicating big changes is the best way to avoid problems in my experience. This has been a pretty easy workflow on this project imo and we've been good about it for the most part. I was just thinking about my commit all week knowing I wouldn't be able to get to it, and thankfully it only broke a couple things, so it's not that bad.

Also the last project I worked on, everyone insisted that they have their own fork, which meant that I as the main dev had to spend all my time doing (enormous) merges and fixing bugs instead of the core code, which is why I'm pretty opposed to everyone having their own branch, lol.

**Reply by Morbidous** [reply]
my current process is to use small branches that describe the feature/bugfix (everything I'm doing right now though is just basic refactoring trying to understand the code).

if master is updated while working, then you rebase your branch onto master (rebase basically means move/copy; rebase onto master moves (really copies) your branch to branch off of the most recent commit on master instead of wherever it really started; if there are conflicts they will have to be resolved in the rebase). then to `merge --no-ff` into master when the work is done. doing this keeps all branches separate from each other over time, and puts the work that goes into a particular feature into it's own section separate from other commits.

but there seems to still be the issue that whoever merges first "wins" and the other person is left with the merge conflicts...

**Comment by jamieyello** [comment]
I also think we can safely delete the entire MonoGame project for now, I don't see how that can fit into any of this any time soon.

**Reply by Morbidous** [reply]
I still haven't gotten to where I quite grok the system, and I have not enough time to work on it (wish I was on vacation this time). For me, MonoGame I know pretty well, and I haven't even heard of Godot before this. But I'll just have to figure it out.

**Reply by jamieyello** [reply]
MonoGame is very nice in that it is very simple, unfortunately it is also very simple. 😆

You will spend months working on a custom solution only to realize someone did it better in Unity/Godot, and all you had to do was spend a weekend learning about it. I used to look at people who made games in Unity in a week with so much jealousy, because in MonoGame everything just takes longer to make. If you know how a game engine works in and out, then you are a force to be reckoned with.

Those are just my thoughts on it, I'm not even saying that I'm right, I just don't know how it's going to fit into the project at this moment.

**Reply by Morbidous** [reply]
I used to code in C++ DirectDraw/Direct3D so XNA/MonoGame was a huge upgrade compared to that. (Same with WinForms, I used to program in raw Win32 C++ so WinForms is just amazing in comparison.) I'm not hugely interested in Unity since I always prefer doing things myself (sometimes to my own detriment; but I have enough successes doing that that I know it's at least sometimes the right approach).

**Comment by jamieyello** [comment]
Why on Earth is our JsonConvert so ancient, lol.

(not touching anything until you're done with restructure)

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/54d4c363-3b58-4c56-b438-167a9d9327d7)

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/e75d693d-61e1-4214-9558-338a308c0ee2)

**Reply by Morbidous** [reply]
ok I will do a bunch of moving around soon then.

**Reply by jamieyello** [reply]
Notice how there are two .sln files? One for Godot, and one in the outer folder. How do you feel about getting rid of the one on the outside? That way me and you always see the same thing. It seems like you generally use the one on the outside and I use the other, so you might not even be aware of some work that I've done, and we have to keep them both up to date. The Godot sln can be opened directly without Godot.

It might not be perfect structurally, but Godot expects the .sln file to be there, and there are no Godot project settings that allow you to move it.

**Reply by Morbidous** [reply]
hmm, I had noticed that a while back and found it annoying yes. 8)

yes there should be only one solution file.

I don't know godot but I don't see why it would need a solution file.

I'll try to read up on it.

**Reply by jamieyello** [reply]
https://github.com/godotengine/godot-proposals/issues/1575

Apparently they added this functionality, I'll look into it.

**Reply by Morbidous** [reply]
gonna take me a while to get up to speed on godot, so I'm even more behind now oh well.

What I would want to do is: separate godot solution, project (UI Library), and program. I have no idea yet if this is possible.

- `Mnes\Solution` would theoretically be the solution folder (I prefer this over having the root folder hold all the solution stuff, but it's not that important, and it also involves more file moves and breaks references etc. so not sure we should).
- `Mnes\Mnes.Ui.Godot` should be the godot UI library (not program)
- and then something like `Mnes\Mnes.Godot` or `Mnes\Mnes.Godot.Program` would be the program project (if that is even a concept in godot which I'm not sure of yet)

I think `project.godot` needs to go in the solution folder. Other than that I need to learn about how godot resources work, paths, file types, etc.

This will take me some time so probably you shouldn't wait for me to do anything more right now since I have to learn this. Maybe what I will do in the future is try contacting you before pushing.

It might take me a week or more to learn this since it's already end of sunday and there's work tomorrow.

FYI my days off will be changing soon from Sat/Sun to... Tue/Wed I think.

**Reply by jamieyello** [reply]
I don't think we need a .sln file for Godot, I think we should be able to just point it towards the main one in project settings.

Looking at them both through a text editor, it's not like the Godot one is special in any way anyway. They both just serve the same purpose of pointing to all the .csproj files.

**Comment by jamieyello** [comment]
These are all pretty outdated, I'll bump all the versions if there is no reason for it

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/9e1f30cd-bce3-428b-9e8c-462fe024f8a6)

**Reply by Morbidous** [reply]
When I tried loading the godot project in godot I noticed it seemed to be complaining about the versions so it's possible that it isn't reading the `Directory.Packages.props` file but I'm not sure. I don't know what that image is from but AutoFixture for example should be set specifically to `4.18.1` already. Actually maybe it's because there are two `.sln` files and the second one is not using `Directory.Packages.props` yet due to the location.

**Reply by jamieyello** [reply]
I think what I'm doing is specifying the version to use in the csproj files https://github.com/MNES-Emulator/MNES/commit/aefab16f2cc9021803e31cdae40ce92f3aa192bc , you probably removed those intentionally.

Yeah it is probably being ignored because of the dual .sln issue. I'll try to get rid of the Godot sln so this won't happen.

**Reply by Morbidous** [reply]
`Directory.Packages.props` lets you set each package's version a single time in a central place rather than having version numbers in dozens or hundreds of .csproj files. But msbuild will search upward stopping at the solution file? (actually not sure) so for the godot solution, it probably doesn't get to the outer root folder where the props files are. This might also be a problem with the `Directory.Build.props` file too.

**Reply by jamieyello** [reply]
I removed the Godot .sln and fixed it so Godot's editor knows where the other one is https://github.com/MNES-Emulator/MNES/commit/2fe68b3ad83120337b1123052d382f20b5d9fc1b

Now we only have one.

**Reply by jamieyello** [reply]
Btw, if you ever need to find it, here's where I found the setting (Project Settings with Advanced Options checked)

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/aca8bc49-9fe5-491f-b6fd-fee1e582ec14)

**Reply by jamieyello** [reply]
Also removed the versioning from .cjproj files https://github.com/MNES-Emulator/MNES/commit/c7259e64469b2e1b44591622c4a5ba33c8d72cc6

**Comment by jamieyello** [comment]
Godot project is now unhappy for some reason

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/d4d2e3cb-df1e-46a9-a51f-54f3ba24e251)

**Reply by jamieyello** [reply]
Looks like it's not referencing the necessary Godot nuget packages correctly anymore

**Reply by jamieyello** [reply]
This is probably relevant

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/de7cc0fa-75f6-4da6-9524-fd3033348294)

**Reply by jamieyello** [reply]
But I don't see any package references in the csproj, I think these have something to do with it

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/eebae146-539d-4291-a9e6-33f595de561b)

**Reply by jamieyello** [reply]
My guess is that something here should be migrated to `Directory.Packages.props`, but I'm not too sure

**Reply by Morbidous** [reply]
Sorry for causing you more work.

Where do you get these errors? Is it in VS?

**Reply by Morbidous** [reply]
I think the godot sdk import has implicit package references that include explicit version numbers and you're supposed to put the version in the props file. Myself I am not getting any build errors at the moment. In the worst case we just abandon the props file if it is not compatible with godot.

**Reply by jamieyello** [reply]
Oh, I am 100% a masochist, don't worry

Yeah VS. I think it's because the SDK defines a version in the .csproj. I imagine the SDK has to be declared in the csproj and the version in the .props file.

**Reply by jamieyello** [reply]
Yeah what you said

**Reply by jamieyello** [reply]
This fixed it, it just disabled it for the Godot project specifically. It would obviously be cleaner to do it the way you said, I don't know how to do that, but I'll mess around with it to try to make that work.

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/5d757983-d48c-4b18-97b1-c29204602bff)

**Reply by jamieyello** [reply]
Yeah this also made it unhappy, have to know more about how SDKs work to fix

![image](https://github.com/MNES-Emulator/MNES/assets/10054829/faae3b13-85de-45b4-8fdc-ce9fe40e910e)

**Reply by jamieyello** [reply]
The SDK brings in all these nuget packages, and that's what makes VS unhappy.

**Reply by Morbidous** [reply]
I've pushed a change that disables the implicit references and adds explicit references with the versions coming from the props file. I'm not 100% sure the version properties are loaded or not in the props file though.

**Reply by Morbidous** [reply]
Do you mean it liked the slash and not the attribute or vice-versa?

**Reply by jamieyello** [reply]
Yeah it likes the slash

**Reply by jamieyello** [reply]
> I've pushed a change that disables the implicit references and adds explicit references with the versions coming from the props file. I'm not 100% sure the version properties are loaded or not in the props file though.

Wow, I have no idea how you managed to figure that out. How did you know about that setting? (also posted [here](https://github.com/MNES-Emulator/MNES/commit/e9f882b149dffa3766dd74e0f6c86d6c2f97b5b6#commitcomment-141454235))
